test/e2e/network/dual_stack.go

```diff
@@ -0,0 +1,299 @@
+		}
+	})
+
+	ginkgo.It("should have ipv4 and ipv6 node podCIDRs", func() {
+		// TODO (aramase) can switch to new function to get all nodes
+		nodeList := framework.GetReadySchedulableNodesOrDie(cs)
+
+		for _, node := range nodeList.Items {
+			framework.ExpectEqual(len(node.Spec.PodCIDRs), 2)
+			// assert podCIDR is same as podCIDRs[0]
+			framework.ExpectEqual(node.Spec.PodCIDR, node.Spec.PodCIDRs[0])
+			// assert one is ipv4 and other is ipv6
+			framework.ExpectEqual(isIPv4CIDR(node.Spec.PodCIDRs[0]) != isIPv4CIDR(node.Spec.PodCIDRs[1]), true)
+		}
+	})
+
+	ginkgo.It("should create pod, add ipv6 and ipv4 ip to pod ips", func() {
+		podName := "pod-dualstack-ips"
+
+		pod := &v1.Pod{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:   podName,
+				Labels: map[string]string{"test": "dualstack-pod-ips"},
+			},
+			Spec: v1.PodSpec{
+				Containers: []v1.Container{
+					{
+						Name:  "dualstack-pod-ips",
+						Image: imageutils.GetE2EImage(imageutils.BusyBox),
+					},
+				},
+			},
+		}
+
+		ginkgo.By("submitting the pod to kubernetes")
+		podClient.CreateSync(pod)
+		framework.ExpectNoError(f.WaitForPodRunning(pod.Name))
+
+		p, err := podClient.Get(pod.Name, metav1.GetOptions{})
+		framework.ExpectNoError(err, "Failed to get pod %q", pod.Name)
+
+		gomega.Expect(p.Status.PodIP).ShouldNot(gomega.BeEquivalentTo(""))
+		gomega.Expect(p.Status.PodIPs).ShouldNot(gomega.BeNil())
+
+		// validate there are 2 ips in podIPs
+		framework.ExpectEqual(len(p.Status.PodIPs), 2)
+		// validate first ip in PodIPs is same as PodIP
+		framework.ExpectEqual(p.Status.PodIP, p.Status.PodIPs[0].IP)
+		// assert 2 pod ips belong to different families
+		framework.ExpectEqual(isIPv4(p.Status.PodIPs[0].IP) != isIPv4(p.Status.PodIPs[1].IP), true)
+
+		ginkgo.By("deleting the pod")
+		err = podClient.Delete(pod.Name, metav1.NewDeleteOptions(30))
+		framework.ExpectNoError(err, "failed to delete pod")
+	})
+
+	// takes close to 140s to complete, so doesn't need to be marked [SLOW]
+	// this test is tagged with phase2 so we can skip this until phase 2 is completed and merged
+	// TODO (aramase) remove phase 2 tag once phase 2 of dual stack is merged
+	ginkgo.It("should be able to reach pod on ipv4 and ipv6 ip [Feature:IPv6DualStackAlphaFeature:Phase2]", func() {
+		serverDeploymentName := "dualstack-server"
+		clientDeploymentName := "dualstack-client"
+
+		// get all schedulable nodes to determine the number of replicas for pods
+		// this is to ensure connectivity from all nodes on cluster
+		nodeList := framework.GetReadySchedulableNodesOrDie(cs)
+		gomega.Expect(nodeList).NotTo(gomega.BeNil())
+
+		if len(nodeList.Items) < 1 {
+			e2elog.Failf("Expect at least 1 node, got %v", len(nodeList.Items))
+		}
+
+		replicas := int32(len(nodeList.Items))
+
+		// to ensure all the pods land on different nodes and we can thereby
+		// validate connectivity across all nodes.
+		serverDeploymentSpec := &apps.Deployment{
```

Comment:
these specs are huge with only subtle differences; consider extracting into a function that generates fixtures